### ISSUE TITLE
Consider letting the transport own the creation of IDuplexPipe

### ISSUE THREAD

**Issue by davidfowl** [aspnet/KestrelHttpServer#2429]
Today the transport is responsible for exposing properties on the connection that can be used by Kestrel to create a duplex pipe. It does this today for a few reasons:

- It wants to directly apply the MaxResponseBufferSize and MaxRequestBuffer size to pipe options
- It communicates "fin" via cancelling the transport's output reader (this is super hacky and we should get rid of it anyways).
- It wants to control where "user code" runs. This includes:
  - ConnectionAdapters (soon to be deprecated)
  - CancellationToken callbacks
  - the request processing delegate

Delegating to the transport has some other benefits:
- Transports aren't tied to a specific pipe implementation.
- We can potentially delete ITransportSchedulerFeature, IApplicationTransportFeature 

To make this change though, we need to pass the relevant settings from kestrel to the transport without coupling.

See https://github.com/geoffkizer/KestrelHttpServer/tree/directsockets for an example of the socket transport with a direct pipe implementation.

/cc @tmds @benaadams (only other people that wrote transports 😄)

**Comment by tmds** [comment]
Probably it also means removing `ApplicationSchedulingMode` which is good to avoid double dispatch to the ThreadPool (e.g. Sockets).

I assume existing implementations can just create the two Pipes similar to how Kestrel does it now, and things will work as before.

**Comment by davidfowl** [comment]
> Probably it also means removing ApplicationSchedulingMode which is good to avoid double dispatch to the ThreadPool (e.g. Sockets).

Not necessarily, the application callback (process request) could still be dispatched by kestrel itself.

> I assume existing implementations can just create the two Pipes similar to how Kestrel does it now, and things will work as before.

That's correct.

**Comment by tmds** [comment]
> Not necessarily, the application callback (process request) could still be dispatched by kestrel itself.

Ah, ok. I see the default is to use the ThreadPool:

https://github.com/aspnet/KestrelHttpServer/blob/af636fc8d4870aeba5b892cdcd3994a5d956c475/src/Kestrel.Core/KestrelServer.cs#L86-L88

So Sockets are doing a double dispatch currently.

**Comment by halter73** [comment]
> So Sockets are doing a double dispatch currently.

Yep. That's why we were surprised that it was helpful to dispatch in SocketAsyncEventArgs.Completed on Windows.

**Comment by tmds** [comment]
@halter73 I didn't realize Windows also did a threadpool dispatch on Windows. When you benchmark Sockets with `SchedulingMode.Inline` what difference do you see with the default (ThreadPool)?

**Comment by tmds** [comment]
> It doesn't, we dispatch and it performs much better than running inline

@davidfowl are you sure? In corefx I see the iocp stuff is using `ThreadPoolBoundHandle`, that sounds as if it is using the ThreadPool.

**Comment by halter73** [comment]
I'm sure we dispatch ourselves on Windows in SocketAwaitable.Complete(). ThreadPoolBoundHandle uses the ThreadPool to handle GetQueuedCompletionStatus on completion port threads, but I don't see anywhere it dispatches to worker threads on its own.

**Comment by muratg** [comment]
@davidfowl are you going to be working on this? If not, please punt. Thanks.
